### PR TITLE
[chalf] div, eq, masked_fill, index_put

### PR DESCRIPTION
**Pull request by kshitij12345**
Ref: https://github.com/pytorch/pytorch/issues/74537

### CODE REVIEW

File: aten/src/ATen/AccumulateType.h [hunk 1]
```diff
@@ -68,8 +68,10 @@ template <> struct AccumulateType<int64_t, true> { using type = int64_t; };
 template <> struct AccumulateType<bool, true> {using type = bool; };
 template <> struct AccumulateType<Half, false> { using type = float; };
 template <> struct AccumulateType<BFloat16, false> { using type = float; };
+template <> struct AccumulateType<c10::complex<Half>, false> { using type = c10::complex<float>; };
 template <> struct AccumulateType<c10::complex<float>, false> { using type = c10::complex<double>; };
 template <> struct AccumulateType<c10::complex<double>, false> { using type = c10::complex<double>; };
+template <> struct AccumulateType<c10::complex<Half>, true> { using type = c10::complex<float>; };
```
Comment (kshitij12345) [inline aten/src/ATen/AccumulateType.h:74]: required by
https://github.com/pytorch/pytorch/blob/a0cc38eaca46e25f561adf9d5b5655f625c39377/aten/src/ATen/native/cuda/Indexing.cu#L54

which is required by `index_put_with_sort_kernel`

File: c10/util/Half.h [hunk 2]
```diff
@@ -448,6 +448,17 @@ struct alignas(4) complex<Half> {
     imag_ = a * d + b * c;
     return *this;
   }
+
+  C10_HOST_DEVICE complex<Half>& operator/=(const complex<Half>& other) {
```
Comment (kshitij12345) [inline c10/util/Half.h:452]: Copied from
https://github.com/pytorch/pytorch/blob/a0cc38eaca46e25f561adf9d5b5655f625c39377/c10/util/complex.h#L247-L258

File: aten/src/ATen/Dispatch.h [hunk 3]
```diff
@@ -416,6 +416,46 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
     }                                                                       \
   }()
 
+#define AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND3(                          \
```
Comment (anjali411) [inline aten/src/ATen/Dispatch.h:419]: One of the ways to measure if the complex half support is stable could be to give complex half first class status and include it in complex types by default and ensure everything works

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 4]
```diff
@@ -481,18 +481,18 @@ void ge_kernel(TensorIteratorBase& iter) {
...
         });
     });
   } else {
-    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(kBFloat16, kHalf, iter.common_dtype(), "eq_cpu", [&]() {
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(kComplexHalf, kBFloat16, kHalf, iter.common_dtype(), "eq_cpu", [&]() {
       cpu_kernel_vec(
         iter,
         [](scalar_t a, scalar_t b) -> scalar_t {
-          return a == b;
+          return static_cast<scalar_t>(a == b);
```
Comment (anjali411) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:495]: This static cast is not needed I think

File: c10/util/Half.h (hunk 2 above)
Comment (anjali411) [reply c10/util/Half.h:452]: maybe just use opmath_t for consistency?

File: aten/src/ATen/cpu/vec/vec_base.h [hunk 5]
```diff
@@ -538,7 +538,7 @@ struct Vectorized {
     // 1 if the pred is true, otherwise 0.
     Vectorized<T> vector;
     for (int i = 0; i != size(); ++ i) {
-      vector[i] = bool(op(values[i], other.values[i]));
+      vector[i] = static_cast<T>(op(values[i], other.values[i]));
```
Comment (anjali411) [inline aten/src/ATen/cpu/vec/vec_base.h:541]: huh?

Comment (kshitij12345) [reply aten/src/ATen/cpu/vec/vec_base.h:541]: To pacify MSVC,

```
runner\_work\pytorch\pytorch\build\aten\src\ATen\native\cpu\BinaryOpsKernel.cpp.DEFAULT.cpp
2022-05-14T09:36:24.0841359Z C:/actions-runner/_work/pytorch/pytorch/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp(500): error C2397: conversion from 'bool' to 'scalar_t' requires a narrowing conversion
2022-05-14T09:36:24.0842387Z C:/actions-runner/_work/pytorch/pytorch/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp(491): error C2398: Element '1': conversion from 'bool' to 'const T &' requires a narrowing conversion
2022-05-14T09:36:24.0843023Z         with
2022-05-14T09:36:24.0843317Z         [
2022-05-14T09:36:24.0843578Z             T=float
2022-05-14T09:36:24.0843938Z         ]
2022-05-14T09:36:24.0844601Z C:/actions-runner/_work/pytorch/pytorch/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp(491): error C2398: Element '1': conversion from 'bool' to 'const T &' requires a narrowing conversion
2022-05-14T09:36:24.0845185Z         with
2022-05-14T09:36:24.0845486Z         [
2022-05-14T09:36:24.0845791Z             T=double
2022-05-14T09:36:24.0846061Z         ]
2022-05-14T09:36:24.0846740Z C:/actions-runner/_work/pytorch/pytorch/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp(491): error C2398: Element '1': conversion from 'bool' to 'float' requires a narrowing conversion
2022-05-14T09:36:24.0847493Z Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29337 for x64
```

Ref: https://github.com/pytorch/pytorch/runs/6433697946?check_suite_focus=true#step:6:22091

Comment (kshitij12345) [reply aten/src/ATen/cpu/vec/vec_base.h:541]: Also note, `vector` holds elements of type T and not bool.

File: c10/util/Half.h [hunk 6]
```diff
@@ -427,19 +427,19 @@ struct alignas(4) complex<Half> {
...
 
-  complex<Half>& operator+=(const complex<Half>& other) {
+  C10_HOST_DEVICE complex<Half>& operator+=(const complex<Half>& other) {
     real_ = static_cast<float>(real_) + static_cast<float>(other.real_);
     imag_ = static_cast<float>(imag_) + static_cast<float>(other.imag_);
     return *this;
   }
 
-  complex<Half>& operator-=(const complex<Half>& other) {
+  C10_HOST_DEVICE complex<Half>& operator-=(const complex<Half>& other) {
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply c10/util/Half.h:436]: I thought might as well have them supported on `CUDA`. (I needed previously when the div on CUDA wasn't jiterated.) Would remove it if it doesn't make sense for now.

File: c10/util/Half.h (hunk 2 above)
Comment (kshitij12345) [reply c10/util/Half.h:452]: `opmath_t` lives in `ATen`. AFAIK, `c10` should not include anything from `ATen`.

**Comment by ngimel** [comment]
I have a general question about why we enable complexHalf support on cpu - we are in fact winding down regular Half support on cpu and advise people not to use it (because sometimes it doesn't work, sometimes it's inaccurate, and it's always slow), so cpu support for chalf outside of basic tensor factories and maybe randn to initialize modules is not needed?

File: aten/src/ATen/AccumulateType.h (hunk 1 above)
Comment (ngimel) [reply aten/src/ATen/AccumulateType.h:74]: `accscalar` in `index_put_with_sort_kernel` should be opmath, in fact, can you please change it there, and remove these additions?

**Review by ngimel — COMMENTED** [review]
This is a cool PR, I've left minor comments.

File: aten/src/ATen/cpu/vec/vec_base.h (hunk 5 above)
Comment (kshitij12345) [reply aten/src/ATen/cpu/vec/vec_base.h:541]: I am not really sure. Unfortunately, I don't have a Windows system on me to verify that quickly.

File: aten/src/ATen/AccumulateType.h (hunk 1 above)
Comment (kshitij12345) [reply aten/src/ATen/AccumulateType.h]: Have updated the `index_put_with_sort_kernel` and removed the new entries.

**Comment by kshitij12345** [comment]
* Have removed `operator/=` from `complex<Half>`.  (other operators are used by previous ops like add, sub, mul which are also implemented on CPU, will remove them in seperate PR).
* Have removed CPU support for `div chalf`.
* Have updated `AccumulateType` usage in `Indexing.cu` to `opmath`.

PTAL @ngimel. Thanks!

**Review by anjali411 — APPROVED** [review]
LGTM.

**Pull request closed without merging**